HW0/P5/P5.txt

```diff
@@ -1 +1,18 @@
 # Your answers here.
+1. 255 seconds. Because cashiers can only count 2 bags per second, they need to perform 255 operations, which amount to 255 seonds.
+
+2. Yes. There 8 cashiers. The idea is divide and conquor. Split 256 bags into 8 groups, each of which contains 32 bags. For each group, one can count them in 31 seconds. Then we merge the 8 counted groups and divide them into 4 groups, each of which contains 2 smaller counted groups. It would take 4 cashiers 1 second to finish counting them. Repeat this until we have 1 group in the end. The time it would take is 31 + 1 + 1 + 1 = 34 seconds.
+
+3. Assuming we have infinite people to count. Each recursion level the number of groups halves. Since 2^8 = 32, it takes 8 seconds.
+
+4. Please take a look at P5.png. Nmax = 3000 in my case.
+
+5. From 3, we have calculated that there are 8 levels in total, and each level takes one second to complete. If communication time is taken into account, there will be one additional second when one level goes to another (e.g., 256 -> 128 takes one second to communicate since people can communicate simultaneously). There are in total 8-1 = 7 merging (communicating) operations. As a result, it takes 8 + 7 = 15 seconds in this case.
+
+6. This problem is a little ambiguous. Based on the piazza discussion, I assume that "you" in the problem is not a cashier.
+ 
+1 bag senario: In the divide step, we hand 256 bags to one cashier, and the cashier can further distributes 128 bags to another one. This happen recursively until each cashier has 1 bag. The time it takes in this step is 1 + log_2(256) = 9. Based on 5, communication time takes 7 seconds. However, in this case, since every cashier can only take 1 bag, there is one additional communication at the beginning. And counting time is 8 seconds. So the total time is 9 + 8 + 8 = 25 seconds.
+
+2 bags senario: Since each cashier can have 2 bags, the divide step takes 8 seconds. Everything else is the same as part 5. So it takes 8 + 15 = 23 seconds.
```

Comment:
This divide-and-conquer kind of strategy is probably how you'd actually want to solve a problem like this, but I don't think this is what was meant by "each cashier takes one bag from you". I think that was intended to specifically mean that you (not a cashier) have to individually hand each bag to one cashier - which takes a lot longer than what you've described here (a minimum of 256 seconds, clearly, since just handing out the bags takes 256 seconds)! Similarly, in the two-bags-each case, I think that's supposed to mean that each cashier has to come up to you and take two bags directly from you (which again puts a lower bound of 128 seconds on the total time).